src/main/java/org/wildfly/security/storage/PasswordStorage.java

```diff
@@ -0,0 +1,222 @@
+        return new PasswordStorage(provider, (PasswordStorageSpi) service.newInstance(null), algorithm);
+    }
+
+    /**
+     * Constructor to create PasswordStorage instance
+     * @param provider {@link Provider} of {@link PasswordStorageSpi} instance
+     * @param spi {@link PasswordStorageSpi} instance
+     * @param type JCA type of PasswordStorage
+     */
+    protected PasswordStorage(Provider provider, PasswordStorageSpi spi, String type) {
+        this.provider = provider;
+        this.spi = spi;
+        this.type = type;
+    }
+
+    /**
+     * Check whether given key has an entry associated with.
+     * @param key key to check existence
+     * @param passwordClass password class to check existence
+     * @param <T> {@link Password} class
+     * @return true in case key exist in storage
+     * @throws StorageException when there is a problem with passwordClass cannot be stored.
+     * @throws UnsupportedPasswordClassException when the passwordClass is not supported
+     */
+    public <T extends Password> boolean exists(String key, Class<T> passwordClass) throws StorageException, UnsupportedPasswordClassException {
+        return spi.exists(key, passwordClass);
+    }
+
+    /**
+     * Remove the passwordClass instance under the key from the storage.
+     * @param key - key to use for removal
+     * @param passwordClass - password class to remove from the key
+     * @param <T> - {@link Password} class
+     * @throws StorageException - if key passwordClass combination doesn't exist or key cannot be removed
+     * @throws UnsupportedPasswordClassException - when the passwordClass is not supported
+     */
+    public <T extends Password> void remove(String key, Class<T> passwordClass) throws StorageException, UnsupportedPasswordClassException {
+        spi.remove(key, passwordClass);
+    }
+
+    /**
+     * Initialize Password Storage service with given attributes.
+     * This procedure should set {@link PasswordStorageSpi#initialized} after successful initialization.
+     *
+     * @param attributes attributes to used to pass information to Password Storage service
+     * @throws StorageException if initialization fails due to any reason
+     */
+    public void initialize(Map<String, String> attributes) throws StorageException {
+        spi.initialize(attributes);
+    }
+
+    /**
+     * Retrieve password stored in the storage under the key and of the passwordClass
+     * @param key key to find password
+     * @param passwordClass password class to retrieve from storage
+     * @param <T> {@link Password} class
+     * @return instance of {@link Password} stored in the storage
+     * @throws StorageException if key passwordClass combination doesn't exist
+     * @throws UnsupportedPasswordClassException - when the passwordClass is not supported
+     */
+    public <T extends Password> T retrieve(String key, Class<T> passwordClass) throws StorageException, UnsupportedPasswordClassException {
```

Comment:
Can we switch from key to credentialName?  Within realms we now reference credentials using credentialName.